Sources/OpenCombine/Publishers/Publishers.Filter.swift

```diff
@@ -0,0 +1,180 @@
+                    isIncluded: @escaping (Upstream.Output) throws -> Bool) {
+            self.upstream = upstream
+            self.isIncluded = isIncluded
+        }
+
+        /// This function is called to attach the specified `Subscriber`
+        /// to this `Publisher` by `subscribe(_:)`
+        ///
+        /// - SeeAlso: `subscribe(_:)`
+        /// - Parameters:
+        ///     - subscriber: The subscriber to attach to this `Publisher`.
+        ///                   once attached it can begin to receive values.
+        public func receive<SubscriberType: Subscriber>(subscriber: SubscriberType)
+            where Upstream.Output == SubscriberType.Input,
+                  SubscriberType.Failure == Failure
+        {
+            let filter = _Filter<Upstream, SubscriberType>(downstream: subscriber,
+                                                           description: "TryFilter",
+                                                           isIncluded: isIncluded)
+            upstream.receive(subscriber: filter)
+        }
+    }
+}
+
+extension Publisher {
+
+    /// Republishes all elements that match a provided closure.
+    ///
+    /// - Parameter isIncluded: A closure that takes one element and returns
+    /// a Boolean value indicating whether to republish the element.
+    /// - Returns: A publisher that republishes all elements that satisfy the closure.
+    public func filter(
+        _ isIncluded: @escaping (Self.Output) -> Bool
+    ) -> Publishers.Filter<Self> {
+        return Publishers.Filter(upstream: self, isIncluded: isIncluded)
+    }
+
+    /// Republishes all elements that match a provided error-throwing closure.
+    ///
+    /// If the `isIncluded` closure throws an error, the publisher fails with that error.
+    ///
+    /// - Parameter isIncluded:  A closure that takes one element and returns a
+    /// Boolean value indicating whether to republish the element.
+    /// - Returns:  A publisher that republishes all elements that satisfy the closure.
+    public func tryFilter(
+        _ isIncluded: @escaping (Self.Output) throws -> Bool
+    ) -> Publishers.TryFilter<Self> {
+        return Publishers.TryFilter(upstream: self, isIncluded: isIncluded)
+    }
+}
+
+private class _Filter<Upstream: Publisher, Downstream: Subscriber>
+    : OperatorSubscription<Downstream>,
+      Subscriber,
+      CustomStringConvertible,
+      Subscription
+    where Upstream.Output == Downstream.Input
+{
+    typealias Input = Upstream.Output
+    typealias Output = Downstream.Input
+    typealias Failure = Upstream.Failure
+
+    private let _isIncluded: (Input) throws -> Bool
+    private var _demand: Subscribers.Demand = .none
+
+    init(downstream: Downstream,
+         description: String,
+         isIncluded: @escaping (Input) throws -> Bool)
+    {
+        self.description = description
+        self._isIncluded = isIncluded
+        super.init(downstream: downstream)
+    }
+
+    let description: String
```

Comment:
Storing `description` is probably not the way to go. Combine uses two different subclasses for `Filter` and `TryFilter`, which share logic in a common base class. Subclassing also allows us to avoid force casts below.

You can refer to `Map`/`TryMap` implementation to see how we do this.